### ISSUE TITLE
Rule proposal: no-assertion-functions

### ISSUE THREAD

**Issue by jendrikw** [typescript-eslint/typescript-eslint#11944]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+new+plugin+rule%22) and found none that match my proposal.
- [x] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [x] My proposal specifically checks TypeScript syntax, or it proposes a check that requires type information to be accurate.
- [x] My proposal is not a "formatting rule"; meaning it does not just enforce how code is formatted (whitespace, brace placement, etc).
- [x] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

Typescripts [assertion functions](https://devblogs.microsoft.com/typescript/announcing-typescript-3-7/#assertion-functions) are very unsound, and I want to forbid them.

### Fail Cases

```TypeScript
function f(x: number | string): asserts x is string {
/// ...
}

function f(x: number | string): asserts x {
/// ...
}
```

### Pass Cases

```TypeScript
function f(x: number | string): x is string {
/// ...
}
```

### Additional Info

The form `x is string` required the body of the function to actually check.

Here is a playground of unsound code: https://www.typescriptlang.org/play/?#code/GYVwdgxgLglg9mABAQwM6oKYCcoElUDKUWMYA5gBQAeAXImCALYBG2iAPoqsaWQJR00mHKkRVEMUdxLlEAbwCwAKAC+y5RATcxdabw70mrLIgC8iALLIoACwB0WZGAAmcRhT6IA-IgCMiOgAiKkCAbmUhbDxCHnJqPg0tOAAbDDtkuEoqOwAHZGciZBwKAFY+BKVlAHoqxFBIWAQ66joGFjZOPXIBMQkpWLJ5atrERCwMKBAsJCphxDVKpSA

**Comment by JoshuaKGoldberg** [comment]
😬 they may be unsound by nature, but they're a very common usage pattern that many projects get a lot of benefit out of. Similar to `as` assertions, they're not something we can reasonably ban in typescript-eslint.

https://typescript-eslint.io/troubleshooting/faqs/general/#how-can-i-ban-specific-language-feature

Thanks for filing, though! If you want to ban them I'd recommend either writing a `no-restricted-syntax` option, or even a custom rule.

**Comment by jendrikw** [comment]
I already ban all usages of `as` except for `as const` with consistent-type-assertions: never and no-unsafe-type-assertion, but you're right that this can easily implemented via `no-restricted-syntax`.
